src/method/query/response.rs

```diff
@@ -0,0 +1,231 @@
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::ops::Index;
+use std::slice::SliceIndex;
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use crate::Result;
+use crate::Value;
+
+/// A wrapper type around the list of results for the queries that were returned
+/// by the database.
+///
+/// Provides utility functions to access the result of one specific query, or if
+/// needed, all queries at once.
+///
+#[derive(Debug, Clone)]
+pub struct QueryResponse(Vec<QueryResult>);
+
+impl QueryResponse {
+    /// Constructs an empty [`QueryResponse`]
+    pub fn new() -> Self {
+        Self(Default::default())
+    }
+
+    /// Returns a reference the result for the `n`-th query from the response. If
+    /// no result is found at this index then [None] is returned.
+    pub fn get_query(&self, n: usize) -> Option<&QueryResult> {
+        self.0.get(n)
+    }
+
+    /// Returns a reference to the result for the `n`-th query from the response.
+    ///
+    /// # Panics
+    /// Unlike the [`Self::get()`] function no bounding check is performed and may panic if
+    /// `n` is out of bounds.
+    pub fn get_query_unchecked(&self, n: usize) -> &QueryResult {
+        &self.0[n]
+    }
+
+    /// Returns a reference to all of the results from the response.
+    pub fn all(&self) -> &Vec<QueryResult> {
+        &self.0
+    }
+
+    /// Returns the deserialized [`<T>`] from the inner [Value]s over the given
+    /// range or index for the query at `query_index`.
+    ///
+    /// - If no query is found at `query_index` then [`None`] is returned.
+    /// - if `index_or_range` is an index of type [usize] then a single [`Option<T>`]
+    /// is returned.
+    /// - if `index_or_range` is a range then a [`Option<Vec<T>>`] is returned if
+    /// and only if the full range is found inside the inner list.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// # #[derive(Debug, serde::Deserialize)]
+    /// # #[allow(dead_code)]
+    /// # struct User {
+    /// #   id: String,
+    /// #   balance: String
+    /// # }
+    /// #
+    /// # use surrealdb_rs::{Result, Surreal};
+    /// # use surrealdb_rs::net::WsClient;
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<()> {
+    /// # let client = Surreal::<WsClient>::new();
+    /// # let token = String::new();
+    /// let response = client.query("select * from user").await?;
+    ///
+    /// // get the first item from the first query
+    /// let user: Option<User> = response.get(0, 0)?;
+    /// tracing::info!("{user:?}");
+    ///
+    /// // get all items from the first query
+    /// let users: Vec<User> = response.get(0, ..)?.unwrap_or_default();
+    /// tracing::info!("{users:?}");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn get<T, I>(&self, query_index: usize, index_or_range: I) -> Result<Option<T>>
```

Comment:
One thing to note here is that you can use the `get` function to retrieve a single element as well as multiple elements. And you're not guaranteed the index/range you provided pointed to an element(s), hence the `Option`. The `Result` is for the errors on the query while the option is for the retrieval of the elements using the `index_or_range`.

There are also cases where having a query return `None` is valid code that should not be treated the same way as a `Result`. The `Option` here allows one to differentiate errors from successful queries that return no results. And having a `Result` indicate a "no results found" may cause such error to be inadvertently _swallowed_ by the `?` operator

Reply:
> One thing to note here is that you can use the get function to retrieve a single element as well as multiple elements

Yes, I'm aware of this.

> And you're not guaranteed the index/range you provided pointed to an element(s), hence the Option

This is where I'm suggesting we return an error if `query_index` or `index_or_range` is out of bounds.

> The Result is for the errors on the query while the option is for the retrieval of the elements using the index_or_range

Yes, those are the normal errors. We can create a special error kind for out of bounds cases like this. Similar to the `std` library's `WouldBlock` IO error kind.

> There are also cases where having a query return None is valid code that should not be treated the same way as a Result

Yes. That's still possible. `T` in this case is any type the compiler infers from the user's code. It can still be `Option<User>` or `Vec<User>` for example.

Reply:
So i think i understood what you said. I've replaced the `Result<Option<T>>` with a `Result<T> where T: Default`, and if either the `query_index` or the `index_or_range` are out of bounds a `Ok(T::default())` is returned. That way if `T` is supposed to be a `Option<something>` it will be None, and if it's a `Vec` it will be empty. This was an excellent idea!

commit: [962db87](https://github.com/surrealdb/surrealdb.rs/pull/11/commits/962db87418603d689c0e7e815743fecd92106309)